### PR TITLE
change MatrixFreeOperators from Number to VectorType template

### PR DESCRIPTION
**Pull request by davydden**
i am working on another vector type, which I would prefer not to discuss at the moment. Regardless, i think it's good to make `MatrixFreeOperators` template with vector type so that it could probably be re-used later for GPU or any other vectors which could be added.

TODO: 
- [x] update step-37

### CODE REVIEW

**Comment by Rombur** [comment]
Would it be possible, instead of using a template for the vector type, to replace `LinearAlgebra::distributed::Vector<Number>` by `LinearAlgebra::VectorSpaceVector<Number>`.  `LA::d::Vector` and the cuda vector already derived from `VectorSpaceVector`. Then, we just need to make sure that your new vector use the new vector interface.

**Comment by davydden** [comment]
@Rombur good idea! I will update, i already have my vector derived from `VectorSpaceVector`, i just did not think about this option.

**Comment by davydden** [comment]
@Rombur @kronbichler the one problem is that we store diagonal of the matrix-free operator. `VectorSpaceVector` is abstract, right? I will see if I can make it work with hard-coded `p::d::Vector` for diagonal while using `VectorSpaceVector` elsewhere...

**Comment by Rombur** [comment]
Yes, `VectorSpaceVector` is abstract. Could we use a pointer to the vector?

**Comment by bangerth** [comment]
As long as you only need the diagonal *internally*, you can use whatever vector you want for the internal interface. In fact, I think that makes sense because then you *know* what you have and what interface is available. The question I think you are trying to answer is simply what vector you want to accept for the *external* interface, and there indeed the long-term plan is to move to the new vector base class.

**Comment by davydden** [comment]
@bangerth diagonal is used externally, the vector is wrapped into `DiagonalMatrix<>` and then is used in preconditioners.

**Comment by davydden** [comment]
@bangerth also some preconditioners require `Matrix` type to provide
```
void precondition_Jacobi(VectorType &dst,
                              const VectorType &src,
                              const value_type omega) const;
```
so diagonal is also used internally.

File: include/deal.II/matrix_free/operators.h [hunk 1]
```diff
@@ -246,7 +248,7 @@ namespace MatrixFreeOperators
      * order to ensure that the cell loops will be able to access the ghost
      * indices with the correct local indices.
      */
-    void adjust_ghost_range_if_necessary(const LinearAlgebra::distributed::Vector<Number> &vec) const;
+    void adjust_ghost_range_if_necessary(const VectorType &vec) const;
```
Comment (kronbichler) [inline include/deal.II/matrix_free/operators.h]: Do you think that this function makes sense on another vector than `LinearAlgebra::distributed::Vector`? I guess we could also move this to a stand-alone function.

File: include/deal.II/matrix_free/operators.h [hunk 2]
```diff
@@ -872,20 +876,21 @@ namespace MatrixFreeOperators
     // lost
     VectorView<Number> view_src_in(src.local_size(), src.begin());
```
Comment (kronbichler) [inline include/deal.II/matrix_free/operators.h]: Here is what I mean: You need to have access to a (raw) pointer of `src` to make this work. I do not think we can implement this for all vector classes. But we do not have to - we would simply require the user to provide us with an appropriate interface.

File: tests/arpack/step-36_parpack_mf.cc [hunk 3]
```diff
@@ -104,8 +104,8 @@ void test ()
 
   std::vector<LinearAlgebra::distributed::Vector<double> > eigenfunctions;
   std::vector<double>                                      eigenvalues;
-  MatrixFreeOperators::MassOperator<dim,fe_degree, fe_degree+1, 1, double> mass;
-  MatrixFreeOperators::LaplaceOperator<dim,fe_degree, fe_degree+1, 1, double> laplace;
+  MatrixFreeOperators::MassOperator<dim,fe_degree, fe_degree+1, 1> mass;
+  MatrixFreeOperators::LaplaceOperator<dim,fe_degree, fe_degree+1, 1> laplace;
```
Comment (kronbichler) [inline tests/arpack/step-36_parpack_mf.cc]: Nit-picking: Is there a reason why you do rely on the default template argument of `MassOperator`? You don't do so below in the other tests. The others have a template `Number` type but I'd prefer to uniformly set it (at least in tests).

Comment (davydden) [inline tests/arpack/step-36_parpack_mf.cc]: historical. Other tests have `number` everywhere, this one was written with `double` from the beginning. So here I just left it default.

**Comment by kronbichler** [comment]
> @Rombur @kronbichler the one problem is that we store diagonal of the matrix-free operator. VectorSpaceVector is abstract, right? I will see if I can make it work with hard-coded p::d::Vector for diagonal while using VectorSpaceVector elsewhere...

The other problem with using `VectorSpaceVector` is that it forces the user to downcast to a concrete vector class in the `local_apply` functions. An alternative would be to internally convert the vector space vector to a read write vector through magic inside `MatrixFree` according to #2535 or to completely hide the vector in the local apply functions... I'm a bit hesitant to use to many abstractions that make implementing non-usual things less flexible, but some way of abstraction/unification would definitely not hurt.

File: include/deal.II/matrix_free/operators.h (hunk 1 above)
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: yes, i think we can move it out. Either a stand-alone function templated with `VectorType` which receives `MPI::Partitioner` or make it a part of each parallel vector. Probably the former is more flexible.

**Comment by Rombur** [comment]
> The other problem with using VectorSpaceVector is that it forces the user to downcast to a concrete vector class in the local_apply functions. An alternative would be to internally convert the vector space vector to a read write vector through magic inside MatrixFree according to #2535 or to completely hide the vector in the local apply functions... 

I think that hiding the vector is the way to go. If we could hide the loop on the cell in `local_apply` and ask for a functor that needs to be applied on quadrature point, we would have a very simple way to switch between cpu and gpu matrix-free. Ideally, we would also hide `reinit(cell)`, `read_dof_values`, and `distribute_local_to_global`. This way, the user doesn't need to know which vector is used. It only needs to give us his discretized formulation at one quadrature point and we can perform all the loops in the order we want.

**Comment by davydden** [comment]
While discussing possible extensions to CUDA we may as well think if something has to be done with `MatrixData<>::initialize_dof_vector<>()`. This is a templated class with a templated function. AFAIK you can do any partial specialization of the function only. So In order for that to be extendable one would need to add a variant of this function for each vector we would like to use. I think a better way is to ask `MatrixData` for `MPI::Partitioner` and then it would be up to a vector class to use the partitioner for initialization. Why I mention it here, because we have
```
  template <int dim, typename VectorType>
  void
  Base<dim,VectorType>::initialize_dof_vector (VectorType &vec) const
  {
    Assert(data.get() != NULL,
           ExcNotInitialized());
    if (!vec.partitioners_are_compatible(*data->get_dof_info(0).vector_partitioner))
      data->initialize_dof_vector(vec);
    Assert(vec.partitioners_are_globally_compatible(*data->get_dof_info(0).vector_partitioner),
           ExcInternalError());
  }
```
which is also a difficult part to extend to other vector types.

**Comment by masterleinad** [comment]
I am working on a version that supports BlockVectors but it should not be a problem to adapt to this new interface.

**Comment by davydden** [comment]
@kronbichler @Rombur @kalj 

i spent some time trying to have an alternative version of this PR with `VectorSpaceVector` in all `apply`-related places and an additional `DiagonalVector` template parameter to specify what is to be used to store the inverse diagonal elements.

At the moment, however, i was not able to make this work. There numerous problems:
1) `vmult_interface_up` in the `Base` class for GMG where we copy a source vector `LinearAlgebra::distributed::Vector<Number> src_cpy (src);`.
2) `precondition_Jacobi` which would use `DiagonalMatrix<DiagonalVector>->vmult()`.
3) calls of `vec.partitioners_are_compatible()` / `vec.partitioners_are_globally_compatible`/ `src.get_partitioner()` etc as those are not there in `VectorSpaceVector`. 
4) calls of type `dst.local_element(constrained_dofs[i]) = 1.;` where we apply constrains. Again, `VectorSpaceVector` does not have `local_element`.

(2) could be hacked to work as long as the provided `VectorSpaceVector` can be multiplied with inverse diagonal stored in `DiagonalVector`. But (1), (3-4) are more fundamental, IMO.
Disclaimer: there  could be more issues, those above are only first errors I got when trying to pursue this direction.

**One could probably make all that work one day, but at the moment I am inclined to think that it does not worth it**. At the end of the day we conceptually consider `MatrixFreeOperators::Base` as a substitute for a matrix to be one-to-one paired with some `VectorType`. Additional benefit to have `VectorType` as a template argument is that we can do some type-traits later to associate GPU-specific matrix-free to CUDA vector.

Bottom line is i would follow the direction of this PR with a `VectorType` template used in all `vmult` related things and to store the inverse diagonal of the operator.

**Comment by Rombur** [comment]
@davydden The interface of `VectorSpaceVector` can be changed if we need it. Actually #3957 adds (1) and has a work-around for (4). I also think it makes sense to add (3). So if you need it, I think we could merge #3957 and then, I can add (3).

**Comment by davydden** [comment]
@Rombur 

> Actually #3957 adds (1) 

does it? There is `reinit()`, but here we create a full copy of an input vector. I am not sure if you can copy a vector based on the parent class only. Unless you do whole lot of dynamic casts to check which vector do you actually have.

> I also think it makes sense to add (3)

That I agree.

---

I think the main question to me so far: what are the advantages to use `VectorSpaceVector` in `MatrixFreeOperators::Base`? In what circumstances would we use such object with different vectors and how is it advantageous to having `DerivedOperator<Vector1>` and `DerivedOperator<Vector2>` which would use the same definitions for `apply_add`? Also note that when we use `VectorSpaceVector`, then in addition to the `Number` template we have to add another `DiagonalVector` template. So the number of template arguments will increase.

**Comment by davydden** [comment]
@Rombur @kronbichler if there are no further comments on the template change in `MatrixFreeOperators`, can we agree on the proposed approach?

**Comment by kronbichler** [comment]
I agree on proceeding with the proposed approach. We definitely need to think more about the interfaces and should target a bigger change for the 9.0 release according to #4011.

**Comment by Rombur** [comment]
Yes, it is fine with me.
